Using armeria grpc client in spring mvc can't propagate trace context

We have a spring-mvc application. It uses spring-cloud-sleuth for distributed tracking. We also use Ameria's Client to call our backend Ameria Grpc service.

Expect:
Tracing should work good, Ameria's client should use the same trace of spring mvc's request.

But in fact:
Ameria's client create a new trace.

More info:

some code:
```java
    @Bean
    RecommendServiceGrpc.RecommendServiceBlockingStub recommendServiceBlockingStub(
            Tracing tracing
    ) {

        return new ClientBuilder("gproto+http://group:ai-recommend-v2/")
                .decorator(HttpTracingClient.newDecorator(tracing))
                .build(RecommendServiceGrpc.RecommendServiceBlockingStub.class);
    }
    @Bean
    CurrentTraceContext.Builder sleuthCurrentTraceContextBuilder() {
        return RequestContextCurrentTraceContext.newBuilder();
    }
```

warning in log:
```
2019-05-14 15:35:50.425  WARN [,,,] 34541 --- [nio-8081-exec-1] .a.c.t.RequestContextCurrentTraceContext : Attempted to propagate trace context, but no request context available. Did you forget to use RequestContext.contextAwareExecutor() or RequestContext.makeContextAware()?
```

We tried to customize the executor of tomcat, but the application would throw RequestContext unavailable when starting up.
```java
    @Bean
    WebServerFactoryCustomizer<TomcatServletWebServerFactory> tomcatServletWebServerFactoryWebServerFactoryCustomizer() {
        return new WebServerFactoryCustomizer<TomcatServletWebServerFactory>() {
            @Override
            public void customize(TomcatServletWebServerFactory factory) {
                factory.addConnectorCustomizers(
                        new TomcatConnectorCustomizer() {
                            @Override
                            public void customize(Connector connector) {
                                ProtocolHandler handler = connector.getProtocolHandler();
                                if (handler instanceof AbstractProtocol) {
                                    AbstractProtocol protocol = (AbstractProtocol) handler;
                                    protocol.setExecutor(
                                            RequestContext.current().makeContextAware(
                                                    protocol.getExecutor()
                                            )
                                    );

                                }
                            }
                        }
                );

            }
        };
    }
```

```
org.springframework.context.ApplicationContextException: Unable to start web server; nested exception is java.lang.IllegalStateException: RequestContext unavailable
.....
Caused by: java.lang.IllegalStateException: RequestContext unavailable
	at com.linecorp.armeria.common.RequestContext.current(RequestContext.java:75) 
```

--- a/zipkin/src/main/java/com/linecorp/armeria/client/tracing/HttpTracingClient.java
+++ b/zipkin/src/main/java/com/linecorp/armeria/client/tracing/HttpTracingClient.java
@@ -25,6 +25,9 @@ import java.util.function.Function;
 
 import javax.annotation.Nullable;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import com.linecorp.armeria.client.Client;
 import com.linecorp.armeria.client.ClientRequestContext;
 import com.linecorp.armeria.client.SimpleDecoratingClient;
@@ -54,6 +57,8 @@ import brave.propagation.TraceContext;
  */
 public class HttpTracingClient extends SimpleDecoratingClient<HttpRequest, HttpResponse> {
 
+    private static final Logger logger = LoggerFactory.getLogger(HttpTracingClient.class);
+
     /**
      * Creates a new tracing {@link Client} decorator using the specified {@link Tracing} instance.
      */
@@ -68,7 +73,13 @@ public class HttpTracingClient extends SimpleDecoratingClient<HttpRequest, HttpR
     public static Function<Client<HttpRequest, HttpResponse>, HttpTracingClient> newDecorator(
             Tracing tracing,
             @Nullable String remoteServiceName) {
-        ensureScopeUsesRequestContext(tracing);
+        try {
+            ensureScopeUsesRequestContext(tracing);
+        } catch (IllegalStateException e) {
+            logger.warn("{} - it is appropriate to ignore this warning if this client is not being used " +
+                        "inside an Armeria server (e.g., this is a normal spring-mvc tomcat server).",
+                        e.getMessage());
+        }
         return delegate -> new HttpTracingClient(delegate, tracing, remoteServiceName);
     }
 

--- a/zipkin/src/test/java/com/linecorp/armeria/client/tracing/HttpTracingClientTest.java
+++ b/zipkin/src/test/java/com/linecorp/armeria/client/tracing/HttpTracingClientTest.java
@@ -17,7 +17,6 @@
 package com.linecorp.armeria.client.tracing;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.argThat;
 import static org.mockito.ArgumentMatchers.same;
@@ -67,12 +66,8 @@ public class HttpTracingClientTest {
     }
 
     @Test
-    public void newDecorator_shouldFailFastWhenRequestContextCurrentTraceContextNotConfigured() {
-        assertThatThrownBy(() -> HttpTracingClient.newDecorator(Tracing.newBuilder().build()))
-                .isInstanceOf(IllegalStateException.class).hasMessage(
-                "Tracing.currentTraceContext is not a RequestContextCurrentTraceContext scope. " +
-                "Please call Tracing.Builder.currentTraceContext(RequestContextCurrentTraceContext.INSTANCE)."
-        );
+    public void newDecorator_shouldWorkWhenRequestContextCurrentTraceContextNotConfigured() {
+        HttpTracingClient.newDecorator(Tracing.newBuilder().build());
     }
 
     @Test

